### PR TITLE
Add exhaustive tests for calendar round tripping and `from_fields` and fix overflow bugs

### PR DESCRIPTION
**Pull request by robertbastian**
This tests the `RataDie -> Date<C> -> Date<Iso> -> Date<C> -> RataDie` round trip for all dates between `-270000-01-01` and `270000-12-31`. It also asserts that the produced `Date<C>` follows the same ordering as the associated `RataDie` (some calendars already test this for some ranges). For this purpose I added the `DateInner: PartialOrd` bound to `Calendar`. Four calendars, (including `AnyCalendar`, which is why it can't be `Ord`) were lacking this implementation.

This discovered a bug in `Hijri`, where the start of the year was stored relative to the approximate `f64` mean synodic year length, not the exact tabular year length (this was missed in #7057). I changed the `debug_assertion!`s in the Hijri code to be actual errors, which are unwrapped in const[^1] for hardcoded data, and can be shown to not happen at runtime because the only year infos we create are tabular, which obeys all the invariants (the simulated rules now fall back to tabular for years where they would break invariants, I couldn't care less about this calendar). These invariants also let us remove the while loops added in #7057.

This PR also adds an exhaustive test for `from_fields` over the extended years `-270000..=270000`, all month codes, days in `1..=100` and ordinal months in `1..=100`.

Both tests require `release-with-assertions` to run in an appropriate time, but the round trip tests then finishes in seconds. We should figure out some way to run at least that one in CI, as it would supersede a lot of manually written tests for specific calendars and ranges, and provides valuable coverage.

[^1]: This requires MSRV 1.83, which is more than 6 versions old, so I can use

### CODE REVIEW

File: components/calendar/src/cal/hijri.rs [hunk 1]
```diff
@@ -357,14 +361,17 @@ impl Rules for TabularAlgorithm {
     }
 
     fn year_data(&self, extended_year: i32) -> HijriYearData {
+        #[allow(clippy::unwrap_used)] // justified, and proven by exhaustive.rs for years +-270_000
```
Comment (Manishearth) [inline components/calendar/src/cal/hijri.rs]: issue: I would still prefer GIGO here with a debug assertion; I _really_ don't want there to be a chance of Chrome or other clients aborting in this case in a release build.

File: components/calendar/src/cal/hijri.rs [hunk 2]
```diff
@@ -491,22 +499,20 @@ impl From<HijriYearData> for i32 {
     }
 }
 
-const LONG_YEAR_LEN: u16 = 355;
```
Comment (Manishearth) [inline components/calendar/src/cal/hijri.rs:506]: question: why remove these?

File: components/calendar/tests/exhaustive.rs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                            types::MonthCode::new_leap(i).unwrap(),
+                        ]
+                        .into_iter()
+                    })
+                    .map(|m| {
+                        let mut fields = types::DateFields::default();
+                        fields.month_code = Some(m);
+                        fields
+                    })
+                    .chain((1..100).map(|m| {
```
Comment (Manishearth) [inline components/calendar/tests/exhaustive.rs]: thought: I think we can cap days at 50 and months at 20

File: components/calendar/tests/exhaustive.rs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    };
+}
+use for_each_calendar;
+
+#[derive(Debug, Clone)]
+struct LunarChineseYears(Vec<cal::chinese::LunarChineseYearData>);
+
+impl LunarChineseYears {
+    fn new<R: cal::chinese::Rules>(r: R) -> Self {
+        Self((-270_001..=270_000).map(|i| r.year_data(i)).collect())
```
Comment (Manishearth) [inline components/calendar/tests/exhaustive.rs]: praise: clever

File: components/calendar/src/calendar.rs [hunk 5]
```diff
@@ -27,7 +27,7 @@ use core::fmt;
 /// </div>
 pub trait Calendar: crate::cal::scaffold::UnstableSealed {
     /// The internal type used to represent dates
-    type DateInner: Eq + Copy + fmt::Debug;
+    type DateInner: Eq + Copy + PartialOrd + fmt::Debug;
```
Comment (Manishearth) [inline components/calendar/src/calendar.rs:33]: nit: document what the PartialOrd should do

(it's kind of obvious, but things can get murky around weird calendars and cached YearInfos)

File: components/calendar/src/provider/hijri.rs [hunk 6]
```diff
@@ -97,6 +104,7 @@ impl PackedHijriYearInfo {
     }
 
     pub(crate) fn last_day_of_month(self, month: u8) -> u16 {
+        // debug_assert!(1 <= month && month <= 12, "{month}");
```
Comment (Manishearth) [inline components/calendar/src/provider/hijri.rs]: nit: clean up?

File: components/calendar/src/cal/hijri.rs [hunk 7]
```diff
@@ -750,13 +759,9 @@ impl<R: Rules> Calendar for Hijri<R> {
...
 
         if rd < year.start_day() {
-            while rd < year.start_day() {
-                year = self.0.year_data(year.extended_year - 1);
-            }
-        } else {
-            while rd >= year.start_day() + year.last_day_of_month(12) as i64 {
-                year = self.0.year_data(year.extended_year + 1)
-            }
+            year = self.0.year_data(year.extended_year - 1);
```
Comment (Manishearth) [inline components/calendar/src/cal/hijri.rs]: thought: still a little wary about removing the while loop.

File: components/calendar/src/cal/hijri.rs (hunk 1 above)
Comment (Manishearth) [reply components/calendar/src/cal/hijri.rs]: Especially if we plan for our valid range to be larger: https://github.com/unicode-org/icu4x/pull/7062#issuecomment-3391169949

Comment (robertbastian) [reply components/calendar/src/cal/hijri.rs]: I can make an unchecked version for tabular

File: components/calendar/src/cal/hijri.rs (hunk 2 above)
Comment (robertbastian) [reply components/calendar/src/cal/hijri.rs:506]: uh I tried to refactor some of the simulated code and prefer them inline, because then you see that the three cases it checks are continuous values

File: components/calendar/tests/exhaustive.rs (hunk 3 above)
Comment (robertbastian) [reply components/calendar/tests/exhaustive.rs]: cool

File: components/calendar/tests/exhaustive.rs (hunk 4 above)
Comment (robertbastian) [reply components/calendar/tests/exhaustive.rs]: this makes it orders of magnitudes faster, which unfortunately means our simple approximation is not simple enough

File: components/calendar/src/cal/hijri.rs (hunk 1 above)
Comment (Manishearth) [reply components/calendar/src/cal/hijri.rs]: Works!

File: components/calendar/src/calendar.rs (hunk 5 above)
Comment (sffc) [reply components/calendar/src/calendar.rs:33]: Question: Why do we need this bound? It was fine before and if you need the impl in a generic context you can add the bound yourself in userland. I'm fine with it, but since it already has Eq, should this be Ord instead?

File: components/calendar/src/cal/hijri.rs [hunk 8]
```diff
@@ -491,22 +499,20 @@ impl From<HijriYearData> for i32 {
...
+    /// to obtain a [`RataDie`] from a [`Date`] in an arbitrary calendar. `start_day` has
+    /// to be within 5 days of the start of the year of the [`TabularAlgorithm`].
     ///
     /// `month_lengths[n - 1]` is true if the nth month has 30 days, and false otherwise.
-    /// Either 6 or 7 months should have 30 days.
-    pub fn new(extended_year: i32, start_day: RataDie, month_lengths: [bool; 12]) -> Self {
-        Self {
-            packed: PackedHijriYearInfo::new(extended_year, month_lengths, start_day),
+    /// Either 6 or 7 months need to have 30 days.
+    pub fn new(extended_year: i32, start_day: RataDie, month_lengths: [bool; 12]) -> Option<Self> {
```
Comment (sffc) [inline components/calendar/src/cal/hijri.rs]: Nit: If it does validation (which seems good), please make it called `try_new` and (optional but preferred) return a `Result`

File: components/calendar/src/cal/hijri.rs (hunk 1 above)
Comment (sffc) [reply components/calendar/src/cal/hijri.rs]: A crate-private `new_unchecked` that debug-asserts seems fine

File: components/calendar/src/cal/hijri.rs (hunk 8 above)
Comment (robertbastian) [reply components/calendar/src/cal/hijri.rs]: `Option` can be unwrapped in `const`, which `Result` cannot. I also don't want to introduce a custom error type.

File: components/calendar/src/calendar.rs (hunk 5 above)
Comment (robertbastian) [reply components/calendar/src/calendar.rs:33]: It's a reasonable bound we can add so that it doesn't have to be added in userland. Dates are comparable, so our `Date` type should be as well.

It cannot be `Ord` because `AnyCalendar` doesn't have a total order

File: components/calendar/src/cal/hijri.rs (hunk 2 above)
Comment (Manishearth) [reply components/calendar/src/cal/hijri.rs:506]: Fair enough. I find that the constant name also serves as documentation, so perhaps include a short comment in the new hardcoded constants saying what they are? Non blocking.

I know we're not consistent about documenting "obvious" per-calendar internals but it's nice to try.

File: components/calendar/src/calendar.rs (hunk 5 above)
Comment (robertbastian) [reply components/calendar/src/calendar.rs:33]: We decided that it's ok to add `PartialEq` to `AnyDateInner` in https://github.com/unicode-org/icu4x/issues/3469

File: components/calendar/fuzz/fuzz_targets/construction.rs [hunk 9]
```diff
@@ -103,7 +103,19 @@ fuzz_target!(|data: FuzzInput| {
     let mut fields = DateFields::default();
     // Temporal only cares about validity in ±270k. We generously test outside of that.
     // We should error on these dates instead, or otherwise handle them: https://github.com/unicode-org/icu4x/issues/7049
-    fields.extended_year = Some(data.year % (i32::MAX / 16));
+    const INTERNAL_EXTENDED_YEAR_OFFSETS: &[i32] = &[
```
Comment (robertbastian) [inline components/calendar/fuzz/fuzz_targets/construction.rs]: @Manishearth this is what it actually boils down to. it's not that we have unexpected internal overflows, we have very expected internal overflows because sometimes we store dates in terms of another calendar (everything else are bugs that this PR fixes). We can use this to justify an `i32::MAX / 2` limit without using the word "overflow", or just use these as calendar-specific limits.

Comment (Manishearth) [reply components/calendar/fuzz/fuzz_targets/construction.rs]: That works for me. Can we land my PR with the i32::MAX / 16 limit and then independently tweak it? I don't want to need to verify that 100% just yet.

Comment (Manishearth) [reply components/calendar/fuzz/fuzz_targets/construction.rs]: (I do plan to review this today but I'd still be wary of whether or not we have caught all of the overflow issues)

Comment (robertbastian) [reply components/calendar/fuzz/fuzz_targets/construction.rs]: yes I approved your PR

Comment (Manishearth) [reply components/calendar/fuzz/fuzz_targets/construction.rs]: And while that's non blocking I'd like to figure out a way to document this: "do not feed us `year` values outside of this range" is easy, "do not do anything that causes our calendars to throw a range error" is less great. We probably could document approximate ranges if we wanted.

Ultimately it's _fine_ since these are errors not assertions, but it's nice to give guidance to the user too.

Comment (Manishearth) [reply components/calendar/fuzz/fuzz_targets/construction.rs]: Ah, thanks.

File: components/calendar/fuzz/fuzz_targets/construction.rs [hunk 10]
```diff
@@ -128,5 +128,16 @@ fuzz_target!(|data: FuzzInput| {
             fields.month_code = Some(unwrap_or_return!(MonthCode::new_leap(data.month)));
         }
     };
-    let _date = Date::try_from_fields(fields, options, calendar);
+    if let Ok(date) = Date::try_from_fields(fields, options, calendar) {
```
Comment (Manishearth) [inline components/calendar/fuzz/fuzz_targets/construction.rs:116]: praise: thanks, this was on the plan but didn't get around to it.

File: components/calendar/fuzz/fuzz_targets/construction.rs (hunk 9 above)
Comment (Manishearth) [reply components/calendar/fuzz/fuzz_targets/construction.rs]: (Also, to be clear, I never thought these were _unexpected_ internal overflows, but I would rather have defense in depth for them before playing whack-a-mole with individual ones)

File: components/calendar/src/cal/ethiopian.rs [hunk 11]
```diff
@@ -14,11 +14,11 @@ use crate::{types, Calendar, Date, RangeError};
 use calendrical_calculations::rata_die::RataDie;
 use tinystr::tinystr;
 
-/// The number of years the Amete Mihret epoch precedes the Coptic epoch
-const AMETE_MIHRET_OFFSET: i32 = 276;
+/// The number of years the Amete Mihret epoch follows the Coptic epoch
```
Comment (Manishearth) [inline components/calendar/src/cal/ethiopian.rs]: issue: this is somewhat confusing with a negative number. Comment with the exact math somewhere? `a_m_epoch = coptic_epoch + AMETE_MIHRET_OFFSET`?

File: components/calendar/src/cal/ethiopian.rs [hunk 12]
```diff
@@ -74,16 +74,16 @@ impl DateFieldsResolver for Ethiopian {
     #[inline]
     fn year_info_from_era(&self, era: &str, era_year: i32) -> Result<Self::YearInfo, DateError> {
         match (self.era_style(), era) {
-            (EthiopianEraStyle::AmeteMihret, "am") => Ok(era_year - AMETE_MIHRET_OFFSET),
-            (_, "aa") => Ok(era_year - AMETE_ALEM_OFFSET),
+            (EthiopianEraStyle::AmeteMihret, "am") => Ok(era_year + AMETE_MIHRET_OFFSET),
```
Comment (Manishearth) [inline components/calendar/src/cal/ethiopian.rs:80]: thought: we've messed this up before, I'm somewhat worried that not all use sites have been caught. Probably fine, worth a double-check

Comment (Manishearth) [reply components/calendar/src/cal/ethiopian.rs:80]: _looks_ right.

File: components/calendar/src/cal/hijri.rs [hunk 13]
```diff
@@ -358,24 +362,27 @@ impl Rules for TabularAlgorithm {
...
-                    || m == 11
-                        && match self.leap_years {
-                            TabularAlgorithmLeapYears::TypeII => {
-                                (14 + 11 * extended_year).rem_euclid(30) < 11
-                            }
+            1,
+            1,
+            self.epoch.rata_die(),
+        );
+        let month_lengths = core::array::from_fn(|m| {
```
Comment (Manishearth) [inline components/calendar/src/cal/hijri.rs:382]: thought (nb): there are only two possible month_lengths, we could just pick between two const arrays here.

harder to extend to multiple leap_year types if we add them.

File: components/calendar/src/any_calendar.rs [hunk 14]
```diff
@@ -149,6 +149,40 @@ pub enum AnyDateInner {
     Roc(<Roc as Calendar>::DateInner),
 }
 
+impl PartialOrd for AnyDateInner {
```
Comment (Manishearth) [inline components/calendar/src/any_calendar.rs:152]: cc @sffc for impact on his macro PR so he knows this is coming.

File: components/calendar/tests/exhaustive.rs [hunk 15]
```diff
@@ -0,0 +1,190 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use icu_calendar::*;
+
+const LOW: i32 = -270_000;
```
Comment (Manishearth) [inline components/calendar/tests/exhaustive.rs]: nit: potentially cite the Temporal spec for why we chose these numbers

https://tc39.es/proposal-temporal/#sec-temporal-instant-range

File: components/calendar/tests/exhaustive.rs [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+const HIGH: i32 = 270_000;
+
+// Check rd -> date -> iso -> date -> rd for whole range
+#[test]
+#[ignore] // takes about 20 seconds in release-with-assertions
+fn check_round_trip() {
```
Comment (Manishearth) [inline components/calendar/tests/exhaustive.rs:12]: thought: I wonder if we should just have a separate `check-release-assertions` task that only runs in the cronjob CI that runs specific tests.

It just calls `cargo test -p calendar --no-ignore --profile whatever -- check_round_trip` etc

Another way to do it is to have these tests have `release_assertions_only` in their name and run `cargo test --workspace --no-ignore --profile whatever -- release_assertions_only`

File: components/calendar/tests/exhaustive.rs [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+            if year % 50000 == 0 {
+                println!("{} {year:?}", cal.as_calendar().debug_name());
+            }
+            for overflow in [options::Overflow::Constrain, options::Overflow::Reject] {
+                let mut options = options::DateFromFieldsOptions::default();
+                options.overflow = Some(overflow);
+                for mut fields in (1..19)
+                    .flat_map(|i| {
+                        [
+                            types::MonthCode::new_normal(i).unwrap(),
```
Comment (Manishearth) [inline components/calendar/tests/exhaustive.rs:52]: thought: maybe worth supporting ordinal months too

File: components/calendar/tests/exhaustive.rs [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+        ]
+        .map(std::thread::spawn)
+        .into_iter()
+        .for_each(|h| h.join().unwrap());
+    };
+}
+use for_each_calendar;
+
+#[derive(Debug, Clone)]
+struct LunarChineseYears(Vec<cal::chinese::LunarChineseYearData>);
```
Comment (Manishearth) [inline components/calendar/tests/exhaustive.rs:177]: nit: comment explaining that this is a cache.

File: components/calendar/fuzz/fuzz_targets/construction.rs [hunk 19]
```diff
@@ -103,7 +103,19 @@ fuzz_target!(|data: FuzzInput| {
     let mut fields = DateFields::default();
     // Temporal only cares about validity in ±270k. We generously test outside of that.
     // We should error on these dates instead, or otherwise handle them: https://github.com/unicode-org/icu4x/issues/7049
-    fields.extended_year = Some(data.year % (i32::MAX / 16));
+    const INTERNAL_EXTENDED_YEAR_OFFSETS: &[i32] = &[
+        -5776, // Coptic -> Ethiopian Amete Alem offset
+        -276,  // Coptic -> Ethiopian Amete Mihret offset
+        -543,  // Gregorian -> Buddhist offset
+        1911,  // Gregorian -> ROC offset
+        -1,    // Hebrew as it compute the Keviyah for the previous year
```
Comment (Manishearth) [inline components/calendar/fuzz/fuzz_targets/construction.rs]: question: is Chinese just fixed now?

File: components/calendar/fuzz/fuzz_targets/construction.rs [hunk 20]
```diff
@@ -103,7 +103,19 @@ fuzz_target!(|data: FuzzInput| {
...
+    ];
+    if data.year < i32::MIN - INTERNAL_EXTENDED_YEAR_OFFSETS.into_iter().min().unwrap()
```
Comment (Manishearth) [inline components/calendar/fuzz/fuzz_targets/construction.rs]: issue: isn't there a chance for overflow here itself? `saturating_sub`?

File: components/calendar/src/cal/chinese/simple.rs [hunk 21]
```diff
@@ -85,17 +77,33 @@ impl super::LunarChineseYearData {
...
             duration: Milliseconds,
         ) -> LocalMoment {
-            let num_periods = ((rata_die - base_moment.rata_die + 1)
-                * MILLISECONDS_IN_EPHEMERIS_DAY
-                - base_moment.local_milliseconds as i64
-                - 1)
-            .div_euclid(duration.0);
-            base_moment + duration * num_periods
+            let diff_millis = (rata_die - base_moment.rata_die) as i128
+                * MILLISECONDS_IN_EPHEMERIS_DAY as i128
```
Comment (Manishearth) [inline components/calendar/src/cal/chinese/simple.rs]: cc @sffc 

https://github.com/unicode-org/icu4x/pull/7052#pullrequestreview-3317268632

Shane didn't want to do millisecond math in i128 here. I tend to agree, we should make this errory.

Comment (sffc) [reply components/calendar/src/cal/chinese/simple.rs]: The approximation works for 600 million years before overflowing. Debug-erroring outside that range seems like it is WAI? `from_fields` should enforce a bound so that such bounds are not reachable. I thought I reviewed another PR recently from @Manishearth that added such bounds.

File: components/calendar/src/any_calendar.rs (hunk 14 above)
Comment (sffc) [reply components/calendar/src/any_calendar.rs:152]: Thanks. The new macro should generalize to this function.

File: components/calendar/src/date.rs [hunk 22]
```diff
@@ -145,6 +149,13 @@ impl<A: AsCalendar> Date<A> {
     /// and the month as either ordinal or month code. It can constrain out-of-bounds values
     /// and fill in missing fields. See [`DateFromFieldsOptions`] for more information.
     ///
+    /// This function will not accept year/extended_year values that are outside of the range `[-2²⁷, 2²⁷]`,
+    /// regardless of the calendar, instead returning a [`DateError::Range`]. This allows us to to keep
```
Comment (Manishearth) [inline components/calendar/src/date.rs:153]: (I like this framing)

File: components/calendar/src/cal/chinese/simple.rs (hunk 21 above)
Comment (Manishearth) [reply components/calendar/src/cal/chinese/simple.rs]: Yes, this PR is attempting to extend those bounds to ±2³⁰.

With this PR, _all_ calendars work with that wide range, as opposed to most calendars working there except for chinese, dangi, and Hijri (and Hijri is a very simple fix).

File: components/calendar/src/calendar.rs (hunk 5 above)
Comment (sffc) [reply components/calendar/src/calendar.rs:33]: Observation: you are implementing the version of PartialOrd that returns None if the calendars differ. That choice appears consistent with the 2-year-old agreement in https://github.com/unicode-org/icu4x/issues/3469#issuecomment-1624149504.

**Comment by Manishearth** [comment]
Github review functionality has been _really_ buggy these days, often on the "files changed" view if you hit refresh (with nothing in the URL query or fragment) it still shows only a part of the changes. Beware.

File: components/calendar/tests/exhaustive.rs (hunk 17 above)
Comment (robertbastian) [reply components/calendar/tests/exhaustive.rs:52]: I am in lines 63-67

File: components/calendar/fuzz/fuzz_targets/construction.rs (hunk 20 above)
Comment (robertbastian) [reply components/calendar/fuzz/fuzz_targets/construction.rs]: not given the list. I could have written this as +5776, but I want the source for that number to be there

File: components/calendar/src/cal/ethiopian.rs (hunk 11 above)
Comment (robertbastian) [reply components/calendar/src/cal/ethiopian.rs]: I just noticed that we're doing it the other way around in `AbstractGregorian`, and that direction makes a bit more sense, as the ROC offset for example is 1911 (positive). I'll rephrase it

File: components/calendar/src/cal/ethiopian.rs (hunk 12 above)
Comment (robertbastian) [reply components/calendar/src/cal/ethiopian.rs:80]: it's a private const, all call sites are in this file, and I updated all of them

File: components/calendar/src/cal/chinese/simple.rs (hunk 21 above)
Comment (robertbastian) [reply components/calendar/src/cal/chinese/simple.rs]: > Debug-erroring outside that range seems like it is WAI?

Debug panics aren't the only thing that happens, we also get arithmetic panics. I'd like to get to a point where we just don't panic. This requires either this math to use 128 bits, or all calendars to not produce dates that would convert to Chinese dates outside this range.

File: components/calendar/src/cal/chinese/simple.rs [hunk 23]
```diff
@@ -85,17 +77,30 @@ impl super::LunarChineseYearData {
...
-            let num_periods = ((rata_die - base_moment.rata_die + 1)
-                * MILLISECONDS_IN_EPHEMERIS_DAY
-                - base_moment.local_milliseconds as i64
-                - 1)
-            .div_euclid(duration.0);
-            base_moment + duration * num_periods
+            let diff_millis = (rata_die - base_moment.rata_die) * MILLISECONDS_IN_EPHEMERIS_DAY
+                - base_moment.local_milliseconds as i64;
+
+            let num_periods =
```
Comment (Manishearth) [inline components/calendar/src/cal/chinese/simple.rs:89]: nit: ideally, explain each step. There's a -1 here that is now easier to document

This is a preexisting issue that need not be fixed here

File: components/calendar/src/cal/chinese/simple.rs [hunk 24]
```diff
@@ -85,17 +77,30 @@ impl super::LunarChineseYearData {
     /// Stays anchored in the Gregorian calendar, even as the Gregorian calendar drifts
     /// from the seasons in the distant future and distant past.
     pub(super) fn simple(utc_offset: Milliseconds, related_iso: i32) -> LunarChineseYearData {
+        // calculates the largest moment such that moment = base_moment + n * duration <= rata_die
         fn periodic_duration_on_or_before(
```
Comment (sffc) [inline components/calendar/src/cal/chinese/simple.rs:81]: This code changed in a significant enough way that I would like to review it in detail. The last time we changed this code (a week ago), I did catch bugs.

File: components/calendar/src/cal/chinese/simple.rs [hunk 25]
```diff
@@ -85,17 +77,30 @@ impl super::LunarChineseYearData {
     /// Stays anchored in the Gregorian calendar, even as the Gregorian calendar drifts
     /// from the seasons in the distant future and distant past.
     pub(super) fn simple(utc_offset: Milliseconds, related_iso: i32) -> LunarChineseYearData {
+        // calculates the largest moment such that moment = base_moment + n * duration <= rata_die
```
Comment (sffc) [inline components/calendar/src/cal/chinese/simple.rs]:
```suggestion
        /// calculates the largest moment such that moment = base_moment + n * duration lands on rata_die (< rata_die + 1)
```

File: components/calendar/src/cal/chinese/simple.rs [hunk 26]
```diff
@@ -85,17 +77,30 @@ impl super::LunarChineseYearData {
...
+            let num_periods =
+                (diff_millis + MILLISECONDS_IN_EPHEMERIS_DAY - 1).div_euclid(duration.0);
+
+            let millis = base_moment.rata_die.to_i64_date() * MILLISECONDS_IN_EPHEMERIS_DAY
+                + base_moment.local_milliseconds as i64
+                + num_periods * duration.0;
+
+            let rata_die = millis / MILLISECONDS_IN_EPHEMERIS_DAY - (millis < 0) as i64;
+            let local_milliseconds = millis % MILLISECONDS_IN_EPHEMERIS_DAY
+                + (millis < 0) as i64 * MILLISECONDS_IN_EPHEMERIS_DAY;
```
Comment (sffc) [inline components/calendar/src/cal/chinese/simple.rs:100]:
```suggestion
            let rata_die = millis.div_euclid(MILLISECONDS_IN_EPHEMERIS_DAY);
            let local_milliseconds = millis.rem_euclid(MILLISECONDS_IN_EPHEMERIS_DAY);
```

File: components/calendar/src/cal/chinese/simple.rs [hunk 27]
```diff
@@ -85,17 +77,30 @@ impl super::LunarChineseYearData {
...
-            base_moment + duration * num_periods
+            let diff_millis = (rata_die - base_moment.rata_die) * MILLISECONDS_IN_EPHEMERIS_DAY
+                - base_moment.local_milliseconds as i64;
+
+            let num_periods =
+                (diff_millis + MILLISECONDS_IN_EPHEMERIS_DAY - 1).div_euclid(duration.0);
+
+            let millis = base_moment.rata_die.to_i64_date() * MILLISECONDS_IN_EPHEMERIS_DAY
+                + base_moment.local_milliseconds as i64
+                + num_periods * duration.0;
```
Comment (sffc) [inline components/calendar/src/cal/chinese/simple.rs:94]: Nit: I don't see how inlining the `impl core::ops::Add<Milliseconds> for LocalMoment` and `impl core::ops::Mul<i64> for Milliseconds` fixes any overflow bugs. Please un-do the inlining or explain why you need to inline it.

File: components/calendar/src/calendar.rs (hunk 5 above)
Comment (sffc) [reply components/calendar/src/calendar.rs:33]: I would prefer for the PartialOrd change to be in a separate PR (which I would probably approve; I just think it's poor hygiene to add random impls in unrelated PRs). But, if you keep it in this PR, please make sure that it is in the commit message that lands on main.

File: components/calendar/src/cal/hijri.rs (hunk 8 above)
Comment (sffc) [reply components/calendar/src/cal/hijri.rs]: Consider `Result<Self, ()>`

File: components/calendar/src/cal/chinese/simple.rs (hunk 27 above)
Comment (Manishearth) [reply components/calendar/src/cal/chinese/simple.rs:94]: I think it doesn't fix overflow bugs: previously this code did, but then we asked to not upgrade `Milliseconds` to `i128`, so now this leftover refactor doesn't really do anything. I think it's overall cleaner, at least in this function, but I think it's fine to split that out into a separate PR.

File: components/calendar/src/cal/chinese/simple.rs (hunk 26 above)
Comment (robertbastian) [reply components/calendar/src/cal/chinese/simple.rs:100]: this is more optimised than euclid. we know that our divisor is positive, which euclidian division does not

File: components/calendar/src/cal/chinese/simple.rs (hunk 27 above)
Comment (robertbastian) [reply components/calendar/src/cal/chinese/simple.rs:94]: I removed the Mul impl because it's more debuggable to have an overflow panic here than there

**Comment by robertbastian** [comment]
@sffc none of you comments seem blocking, why are you blocking this PR?

File: components/calendar/src/calendar.rs (hunk 5 above)
Comment (robertbastian) [reply components/calendar/src/calendar.rs:33]: #7090

File: components/calendar/src/cal/hijri.rs (hunk 8 above)
Comment (robertbastian) [reply components/calendar/src/cal/hijri.rs]: `Option` can be unwrapped in `const`, which `Result` cannot.

File: components/calendar/src/cal/chinese/simple.rs (hunk 26 above)
Comment (sffc) [inline components/calendar/src/cal/chinese/simple.rs:100]: When the arithmetic code increases in complexity, it should be associated with a comment explaining why we are increasing the complexity

```suggestion
            // Note: this is Euclidean div/rem, but this more optimized, because
            // we know that our divisor is positive
            let rata_die = millis / MILLISECONDS_IN_EPHEMERIS_DAY - (millis < 0) as i64;
            let local_milliseconds = millis % MILLISECONDS_IN_EPHEMERIS_DAY
                + (millis < 0) as i64 * MILLISECONDS_IN_EPHEMERIS_DAY;
```

File: components/calendar/src/cal/chinese/simple.rs (hunk 25 above)
Comment (sffc) [reply components/calendar/src/cal/chinese/simple.rs]: ^ Please fix this; you are adding a comment that is wrong

**Review by sffc — APPROVED** [review]
